### PR TITLE
Test profil page

### PR DESCRIPTION
**Pull request by Showcas**
tested profile page and the the edit pages.

review will come in its own PR

- wasn't able to test the Avatars

We should have now over 80% Coverage

### CODE REVIEW

File: test/util/factories/ride_factory.dart [hunk 1]
```diff
@@ -54,7 +54,7 @@ class RideFactory extends TripFactory<Ride> {
       startTime: startTime ?? DateTime.now(),
       end: end ?? faker.address.city(),
       endPosition: endPosition ?? Position(faker.geo.latitude(), faker.geo.longitude()),
-      endTime: endTime ?? DateTime.now(),
+      endTime: endTime ?? DateTime.now().add(const Duration(hours: 1)),
```
Comment (Showcas) [inline test/util/factories/ride_factory.dart:57]: this will be replaced once PR 107 is merged. I needed it to not have failing tests all the time

File: test/widgets/profile/edit_brith_date_page_test.dart [hunk 2]
```diff
@@ -0,0 +1,140 @@
...
+          aud: 'public',
+          createdAt: DateTime.now().toIso8601String(),
+          email: email),
+      'email': email,
+    });
+    whenRequest(processor, urlMatcher: contains('/rest/v1/profiles')).thenReturnJson(profile.toJsonForApi());
+  });
+  group('edit_birth_date_page', () {
+    testWidgets('birthDate TextField', (WidgetTester tester) async {
+      //load page
```
Comment (Showcas) [inline test/widgets/profile/edit_brith_date_page_test.dart:50]: maybe here feedback from all. Do you like this kind of documentation or should I remove it?

File: .vscode/settings.json [hunk 3]
```diff
@@ -21,5 +21,8 @@
   "files.watcherExclude": {
     "**/.fvm": true
   },
-  "java.configuration.updateBuildConfiguration": "automatic"
+  "java.configuration.updateBuildConfiguration": "automatic",
+  "cSpell.words": [
```
Comment (fynsta) [inline .vscode/settings.json:25]: can we remove this from the PR? you should be able to put it in your own personal `settings.json`, but it does not need to be in the project file.

File: test/widgets/profile/edit_brith_date_page_test.dart (hunk 2 above)
Comment (fynsta) [reply test/widgets/profile/edit_brith_date_page_test.dart:50]: Remove it, please! :D

File: lib/account/pages/edit_account/edit_profile_features_page.dart [hunk 4]
```diff
@@ -32,6 +32,14 @@ class _EditProfileFeaturesPageState extends State<EditProfileFeaturesPage> {
     _otherFeatures = Feature.values.where((Feature e) => !_features.contains(e)).toList();
   }
 
+  ReorderableDragStartListener buildDragStartListener(int index) {
```
Comment (fynsta) [inline lib/account/pages/edit_account/edit_profile_features_page.dart:35]: what's the point of moving this into a method?

File: lib/account/widgets/avatar.dart [hunk 5]
```diff
@@ -59,6 +59,7 @@ class AvatarState extends State<Avatar> {
                   onTap: () => Navigator.of(context).push(
                     MaterialPageRoute<void>(builder: (BuildContext context) => AvatarPicturePage(widget.profile)),
                   ),
+                  key: const Key('AvatarTappable'),
```
Comment (fynsta) [inline lib/account/widgets/avatar.dart:62]: the "a" should not be capital, same for the other key in this file

File: pubspec.lock [hunk 6]
```diff
@@ -42,7 +42,7 @@ packages:
       name: async
       url: "https://pub.dartlang.org"
     source: hosted
-    version: "2.10.0"
```
Comment (fynsta) [inline pubspec.lock:45]: the pubspec is weird? do you have other flutter versions or something?

File: test/util/factories/profile_factory.dart [hunk 7]
```diff
@@ -32,7 +32,10 @@ class ProfileFactory extends ModelFactory<Profile> {
       username: username ?? faker.internet.userName(),
       email: email ?? faker.internet.email(),
       description: getNullableParameterOr(description, faker.lorem.sentences(random.nextInt(2) + 1).join(' ')),
-      birthDate: getNullableParameterOr(birthDate, faker.date.dateTime()),
+      birthDate: getNullableParameterOr(
+          birthDate,
+          faker.date.dateTime(
+              minYear: DateTime(1900).year, maxYear: DateTime.now().subtract(const Duration(days: 365 * 12)).year)),
```
Comment (fynsta) [inline test/util/factories/profile_factory.dart:38]: nice, but please add a trailing comma
(and we should probably save the min age somewhere in the Profile class and only reference it here)

File: test/widgets/profile/edit_brith_date_page_test.dart [hunk 8]
```diff
@@ -0,0 +1,140 @@
...
+      await tester.tap(birthDateInput);
+      await tester.pumpAndSettle();
+      expect(find.byType(DatePickerDialog), findsOneWidget);
+    });
+    testWidgets('change birthDate', (WidgetTester tester) async {
+      //load page
+      await pumpMaterial(tester, EditBirthDatePage(profile));
+      await tester.pump();
+
+      //check if birthDate is displayed
```
Comment (fynsta) [inline test/widgets/profile/edit_brith_date_page_test.dart:71]: didn't you do exactly this in line 54 alraedy? these tests have a little too much overlap

File: test/widgets/profile/edit_brith_date_page_test.dart [hunk 9]
```diff
@@ -0,0 +1,140 @@
...
+      expect(clearButton, findsOneWidget);
+
+      //check if birthDate is cleared
+      await tester.tap(clearButton);
+      await tester.pump();
+      expect(find.text(localeManager.formatDate(profile.birthDate!)), findsNothing);
+    });
+    testWidgets('save Button', (WidgetTester tester) async {
+      //sign in
+      SupabaseManager.supabaseClient.auth.signInWithPassword(
```
Comment (fynsta) [inline test/widgets/profile/edit_brith_date_page_test.dart:114]: let's put this in setUp, it's not immediately clear that this is specific to this one case

File: test/widgets/profile/edit_brith_date_page_test.dart [hunk 10]
```diff
@@ -0,0 +1,140 @@
...
+
+      //check if save Button is displayed
+      expect(find.byKey(const Key('saveButton')), findsOneWidget);
+
+      //tap save Button
+      await tester.tap(find.byKey(const Key('saveButton')));
+      await tester.pumpAndSettle();
+
+      //check if ProfilePage is displayed
+      expect(find.byType(ProfilePage), findsOneWidget);
```
Comment (fynsta) [inline test/widgets/profile/edit_brith_date_page_test.dart:137]: this should test that the actually correct new age is shown now

File: lib/account/pages/edit_account/edit_gender_page.dart [hunk 11]
```diff
@@ -48,6 +48,7 @@ class _EditGenderPageState extends State<EditGenderPage> {
                           _gender = value;
                         });
                       },
+                      key: Key('${index}genderRadioListTile'),
```
Comment (fynsta) [inline lib/account/pages/edit_account/edit_gender_page.dart:51]: the index should be at the end, not the beginning

File: test/widgets/profile/edit_gender_page_test.dart [hunk 12]
```diff
@@ -0,0 +1,117 @@
...
+
+      //check if diverse is displayed
+      final diverseGenderFinder = find.byKey(const Key('2genderRadioListTile'));
+      expect(diverseGenderFinder, findsOneWidget);
+
+      //check if preferNotToSay is displayed
+      final preferNotToSayGenderFinder = find.byKey(const Key('preferNotToSayGenderRadioListTile'));
+      expect(preferNotToSayGenderFinder, findsOneWidget);
+    });
+    // don't know to test what I have selected
```
Comment (fynsta) [inline test/widgets/profile/edit_gender_page_test.dart:67]: ?

Comment (fynsta) [reply test/widgets/profile/edit_gender_page_test.dart:67]: I don't think this test is useful anyways? It just tests that the RadioListTile works, which should be guaranteed by Flutter.

Comment (Showcas) [reply test/widgets/profile/edit_gender_page_test.dart:67]: ok, yea gonna fix that

File: test/widgets/profile/edit_brith_date_page_test.dart (hunk 10 above)
Comment (Showcas) [reply test/widgets/profile/edit_brith_date_page_test.dart:122]: we can't test if the changes from edit are shown in profile, since it fetches the new values from Supabase. So it would just be me telling it in any situation that it should change the value.
The only thing I could do is that I check if an Supabase update was called after the tap on savebutton.
And then on Profile page if loadProfile is called again.

Comment (fynsta) [reply test/widgets/profile/edit_brith_date_page_test.dart:122]: Yes please just test that there is a supabase call to patch the data

File: lib/account/models/profile.dart [hunk 13]
```diff
@@ -51,6 +51,8 @@ class Profile extends Model {
 
   bool get isCurrentUser => id == SupabaseManager.getCurrentProfile()?.id;
 
+  static int get minAge => DateTime(1900).year;
```
Comment (istzen) [inline lib/account/models/profile.dart:54]: shouldn't that be maxAge?

Comment (istzen) [reply lib/account/models/profile.dart:54]: and maybe we should also add minAge, so that if we change that we don't have to change it in the factory

Comment (fynsta) [reply lib/account/models/profile.dart:54]: yes that was actually what I meant? I think Cedric just misunderstood.

File: test/widgets/profile/edit_brith_date_page_test.dart [hunk 14]
```diff
@@ -0,0 +1,136 @@
...
+    await SupabaseManager.supabaseClient.auth.signInWithPassword(
+      email: email,
+      password: authId,
+    );
+    whenRequest(processor, urlMatcher: contains('/rest/v1/profile')).thenReturnJson(profile.toJsonForApi());
+  });
+  group('edit_birth_date_page', () {
+    testWidgets('birthDate TextField', (WidgetTester tester) async {
+      await pumpMaterial(tester, EditBirthDatePage(profile));
+      await tester.pump();
```
Comment (istzen) [inline test/widgets/profile/edit_brith_date_page_test.dart:57]: i think you don't need this line. Also for the other tests

File: test/widgets/profile/edit_brith_date_page_test.dart [hunk 15]
```diff
@@ -0,0 +1,136 @@
...
+
+      final Finder birthDateFinder = find.text(localeManager.formatDate(profile.birthDate!));
+      final Finder birthDateInputFinder = find.byKey(const Key('birthDateInput'));
+
+      expect(birthDateFinder, findsOneWidget);
+      expect(birthDateInputFinder, findsOneWidget);
+    });
```
Comment (istzen) [inline test/widgets/profile/edit_brith_date_page_test.dart:64]: can you also add linebreaks between the tests

File: lib/account/pages/profile_page.dart [hunk 16]
```diff
@@ -271,12 +277,12 @@ class _ProfilePageState extends State<ProfilePage> {
       if (!_profile!.isCurrentUser) {
         final bool hasRecentReport = _profile!.reportsReceived!
             .any((Report report) => report.isRecent && report.reporterId == SupabaseManager.getCurrentProfile()!.id);
-
```
Comment (fynsta) [inline lib/account/pages/profile_page.dart:274]: why do you hate proper spacing between lines? :D

File: lib/account/widgets/editable_row.dart [hunk 17]
```diff
@@ -33,7 +33,10 @@ class EditableRow extends StatelessWidget {
                   child: IconButton(
                     tooltip: S.of(context).edit,
                     icon: const Icon(Icons.edit),
-                    onPressed: onPressed,
+                    onPressed: () {
```
Comment (fynsta) [inline lib/account/widgets/editable_row.dart:36]: what's the point of changing `onPressed`? It worked fine the way it was (and it was shorter too)

File: test/util/mocks/request_processor.mocks.dart [hunk 18]
```diff
@@ -18,8 +18,7 @@ import 'request_processor.dart' as _i2;
 // ignore_for_file: camel_case_types
 // ignore_for_file: subtype_of_sealed_class
 
-class _FakeProcessorResult_0 extends _i1.SmartFake
-    implements _i2.ProcessorResult {
+class _FakeProcessorResult_0 extends _i1.SmartFake implements _i2.ProcessorResult {
```
Comment (fynsta) [inline test/util/mocks/request_processor.mocks.dart:21]: I assume this change happened when you formatted the project? I would revert it because it's an autogenerated file, not our concern/responsibility

File: test/widgets/profile/edit_brith_date_page_test.dart [hunk 19]
```diff
@@ -0,0 +1,136 @@
...
+      verifyRequest(
+        processor,
+        urlMatcher: equals('/rest/v1/profiles?id=eq.1'),
+        methodMatcher: equals('PATCH'),
+        bodyMatcher: equals({'birth_date': null}),
+      ).called(1);
+
+      verifyRequest(processor,
+              urlMatcher: equals(
+                  '/rest/v1/profiles?select=%2A%2Cprofile_features%28%2A%29%2Creviews_received%3Areviews%21reviews_receiver_id_fkey%28%2A%2Cwriter%3Awriter_id%28%2A%29%29%2Creports_received%3Areports%21reports_offender_id_fkey%28%2A%29&id=eq.1'))
```
Comment (fynsta) [inline test/widgets/profile/edit_brith_date_page_test.dart:132]: okay you don't need to match the whole URL. `startsWith('/rest/v1/profiles')` is enough. Or maybe a regex to capture the id at the end, but that would probably be overkill.

File: test/widgets/profile/edit_brith_date_page_test.dart [hunk 20]
```diff
@@ -0,0 +1,136 @@
...
+      final Finder clearButton = find.byKey(const Key('clearButton'));
+      await tester.tap(clearButton);
+      await tester.pump();
+
+      final Finder saveButtonFinder = find.byKey(const Key('saveButton'));
+      expect(saveButtonFinder, findsOneWidget);
+      await tester.tap(saveButtonFinder);
+      await tester.pumpAndSettle();
+
+      expect(find.byType(ProfilePage), findsOneWidget);
```
Comment (istzen) [inline test/widgets/profile/edit_brith_date_page_test.dart:121]: can you also check that the age is not there anymore

File: test/widgets/profile/edit_brith_date_page_test.dart [hunk 21]
```diff
@@ -0,0 +1,136 @@
...
+
+      verifyRequest(
+        processor,
+        urlMatcher: equals('/rest/v1/profiles?id=eq.1'),
+        methodMatcher: equals('PATCH'),
+        bodyMatcher: equals({'birth_date': null}),
+      ).called(1);
+
+      verifyRequest(processor,
```
Comment (istzen) [inline test/widgets/profile/edit_brith_date_page_test.dart:130]: why does this need to be verified. And can you shorten the url request so that just the relevant part is shown and add methodMatcher: equals('GET') so there is no overlap.

File: test/widgets/profile/edit_brith_date_page_test.dart (hunk 20 above)
Comment (fynsta) [reply test/widgets/profile/edit_brith_date_page_test.dart:121]: No you can't because the profile is reloaded at this point. I mean you could fake the response again, but what would be the point?

File: test/widgets/profile/edit_brith_date_page_test.dart (hunk 21 above)
Comment (fynsta) [reply test/widgets/profile/edit_brith_date_page_test.dart:130]: I think it's fine to verify the account is reloaded. On the other points, I agree

File: test/widgets/profile/edit_description_page_test.dart [hunk 22]
```diff
@@ -0,0 +1,115 @@
...
+      await tester.pump();
+
+      expect(find.text(profile.description!), findsOneWidget);
+
+      final Finder descriptionInput = find.byKey(const Key('description'));
+      expect(descriptionInput, findsOneWidget);
+
+      await tester.tap(descriptionInput);
+      await tester.pumpAndSettle();
+      await tester.enterText(descriptionInput, 'newDescription');
```
Comment (istzen) [inline test/widgets/profile/edit_description_page_test.dart:65]: why is this needed?

File: test/widgets/profile/profile_page_test.dart [hunk 23]
```diff
@@ -0,0 +1,410 @@
...
+    for (int i = 0; i < 2; i++) {
+      i == 0
+          ? SupabaseManager.setCurrentProfile(profile)
+          : SupabaseManager.setCurrentProfile(ProfileFactory().generateFake(id: 2));
+      testWidgets('Works with id parameter', (WidgetTester tester) async {
+        await pumpMaterial(tester, ProfilePage(profile.id!));
+        expect(find.byType(CircularProgressIndicator), findsOneWidget);
+
+        await tester.pump();
+        expect(find.text(profile.username), findsNWidgets(2));
```
Comment (istzen) [inline test/widgets/profile/profile_page_test.dart:87]: can you add a comment why the username is shown 2 times?

File: test/widgets/profile/profile_page_test.dart [hunk 24]
```diff
@@ -0,0 +1,410 @@
...
+
+        final scrollableFinder = find.byType(Scrollable).first;
+        expect(scrollableFinder, findsOneWidget);
+
+        for (int h = 0; h < profile.features!.length; h++) {
+          final feature = profile.features![h];
+          final featureFinder = find.byKey(Key('feature_${feature.name}'));
+          await tester.scrollUntilVisible(featureFinder, 500, scrollable: scrollableFinder);
+          expect(featureFinder, findsOneWidget);
+        }
```
Comment (istzen) [inline test/widgets/profile/profile_page_test.dart:156]: can you also check that no other features are shown, i.e. check if the feature count is the profile.features.length

File: lib/account/widgets/editable_row.dart (hunk 17 above)
Comment (Showcas) [reply lib/account/widgets/editable_row.dart:36]: my mistake, didn't change it back after debuging

File: test/widgets/profile/edit_description_page_test.dart (hunk 22 above)
Comment (Showcas) [reply test/widgets/profile/edit_description_page_test.dart:65]: I test if the discription is editable but I jsut noticed its way too long. I'll fix it

File: lib/account/pages/profile_page.dart (hunk 16 above)
Comment (Showcas) [reply lib/account/pages/profile_page.dart:274]: sorry ist bei debugging entstanden

File: test/util/mocks/request_processor.mocks.dart (hunk 18 above)
Comment (Showcas) [reply test/util/mocks/request_processor.mocks.dart:21]: how do I revert that? If I save it formats it back all the time

Comment (fynsta) [reply test/util/mocks/request_processor.mocks.dart:21]: Sorry I should have changed the README. It's
`flutter format --line-length=120 . $(find lib test -name "*.dart" -not -name "*.mocks.dart")`

**Comment by fynsta** [comment]
I merged the TripTimes PR, you can remove your fix and rebase

File: lib/account/models/profile.dart [hunk 25]
```diff
@@ -51,6 +51,9 @@ class Profile extends Model {
 
   bool get isCurrentUser => id == supabaseManager.currentProfile?.id;
 
+  static int get maxAge => DateTime(1900).year;
+  static int get minAge => DateTime.now().subtract(const Duration(days: 365 * 12)).year;
```
Comment (fynsta) [inline lib/account/models/profile.dart:55]: a year is not an age, that makes no sense

Just use `static const int minAge = const Duration(days: 365 * 12)`. As for the oldest possible age, let's just use 120 or something?

Comment (Showcas) [reply lib/account/models/profile.dart]: I changed it to just const ints. Do we really want 12 as the minimum Age. Or should we go for 16 or 18

Comment (fynsta) [reply lib/account/models/profile.dart]: Oh sorry I meant `Duration`, I don't think `int` adds any valuable information?

Also please use the constant in `EditBirthDate` too, that was the whole point imo.

I think 12 is fine (why should they not be allowed?), but maybe we should not allow young people to create drives? Does not really matter too much to me though.

File: test/util/factories/profile_factory.dart [hunk 26]
```diff
@@ -32,7 +32,12 @@ class ProfileFactory extends ModelFactory<Profile> {
       username: username ?? faker.internet.userName(),
       email: email ?? faker.internet.email(),
       description: getNullableParameterOr(description, faker.lorem.sentences(random.nextInt(2) + 1).join(' ')),
-      birthDate: getNullableParameterOr(birthDate, faker.date.dateTime()),
+      birthDate: getNullableParameterOr(
+          birthDate,
+          faker.date.dateTime(
```
Comment (fynsta) [inline test/util/factories/profile_factory.dart:37]: just controlling the year is not enough, this could still result in wrong ages. It would be better to use `dateTimeBetween`.

**Review by fynsta — APPROVED** [review]

**Review by istzen — APPROVED** [review]

File: test/widgets/profile/edit_brith_date_page_test.dart [hunk 27]
```diff
@@ -0,0 +1,138 @@
+import 'package:flutter/material.dart';
```
Comment (Eddie-42) [inline test/widgets/profile/edit_brith_date_page_test.dart:1]: typo in the file name

**Review by Eddie-42 — APPROVED** [review]

**Review by MarkFalke — APPROVED** [review]

**Comment by fynsta** [comment]
Fixed the tests after the rebase (not 100% sure because they seemed flaky, but now they consistently work for me...). Anyways, I will merge this and we will see what happens.
